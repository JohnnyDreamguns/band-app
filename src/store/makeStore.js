import React, { useState, useMemo, useContext } from 'react';

function makeStore(initialState) {
  const context = React.createContext();

  const StoreProvider = ({ initialValue, children }) => {
    const [state, setState] = useState(initialValue || initialState || {});
    const contextValue = useMemo(() => [state, setState], [state]);
    return <context.Provider value={contextValue}>{children}</context.Provider>;
  };

  const useStore = () => {
    const [state, setState] = useContext(context);
    return { state, setState };
  };

  return { StoreProvider, useStore };
}

export { makeStore };
