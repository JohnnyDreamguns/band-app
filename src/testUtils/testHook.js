import React from 'react';
import TestRenderer from 'react-test-renderer';

const TestHook = ({ callback }) => {
  callback();
  return null;
};

export const testHook = (callback, StoreProvider = null) => {
  if (StoreProvider) {
    return TestRenderer.create(
      <StoreProvider>
        <TestHook callback={callback} />
      </StoreProvider>
    );
  } else {
    return TestRenderer.create(<TestHook callback={callback} />);
  }
};
