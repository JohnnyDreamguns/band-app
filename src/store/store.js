import { makeStore } from './makeStore';

const initialValue = {
  data: {
    bands: [],
    albums: []
  },
  homePage: {
    bands: [],
    loading: false,
    page: 1
  },
  bandPage: {
    bandId: undefined,
    loading: false,
    albums: []
  },
  albumPage: {
    albumId: undefined,
    loading: false
  },
  messages: []
};

const { StoreProvider, useStore } = makeStore(initialValue);

export { StoreProvider, useStore };
