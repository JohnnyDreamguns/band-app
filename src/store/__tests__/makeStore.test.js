import React from 'react';
import { makeStore } from '../makeStore';
import ShallowRenderer from 'react-test-renderer/shallow';
import { testHook } from '../../testUtils/testHook';
import { StoreProvider as SP } from '../../store/store';

const mockState = {
  test: 'mock!',
  test2: 'mock!'
};

const renderer = new ShallowRenderer();

const { useStore } = makeStore();

let useStoreHook;
beforeEach(() => {
  testHook(() => {
    useStoreHook = useStore();
  }, SP);
});

describe('makeStore', () => {
  it('context provider with initialState should match snapshot', () => {
    const { StoreProvider } = makeStore(mockState);
    renderer.render(<StoreProvider />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  it('context provider with initialValue should match snapshot', () => {
    const { StoreProvider } = makeStore();
    renderer.render(<StoreProvider initialValue={{ tester: 'sdsdvd' }} />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  it('context provider with no initialState and no initialValue should match snapshot', () => {
    const { StoreProvider } = makeStore();
    renderer.render(<StoreProvider />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  it('useStore hook have a setState function', () => {
    expect(useStoreHook.setState).toEqual(expect.any(Function));
  });

  it('useStore hook should return expected value', () => {
    expect(useStoreHook.state).toStrictEqual({});
  });
});
