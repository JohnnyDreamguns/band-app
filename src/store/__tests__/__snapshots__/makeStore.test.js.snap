// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeStore context provider with initialState should match snapshot 1`] = `
<Context.Provider
  value={
    Array [
      Object {
        "test": "mock!",
        "test2": "mock!",
      },
      [Function],
    ]
  }
/>
`;

exports[`makeStore context provider with initialValue should match snapshot 1`] = `
<Context.Provider
  value={
    Array [
      Object {
        "tester": "sdsdvd",
      },
      [Function],
    ]
  }
/>
`;

exports[`makeStore context provider with no initialState and no initialValue should match snapshot 1`] = `
<Context.Provider
  value={
    Array [
      Object {},
      [Function],
    ]
  }
/>
`;
